### ISSUE TITLE
Entire SIP MESSAGE in content from message event

### ISSUE THREAD

**Issue by tijmenNL** [meetecho/janus-gateway#2663]
We're testing with receiving SIP MESSAGEs on accounts registered through the SIP plugin. 

If you receive a couple of messages really short after each other, sometimes the content field of the message event from Janus will contain the parsed body of the message **and** the next entire SIP Message. I don't see this coming in on the network interface. There they are normal separated messages.

See this gist:
https://gist.github.com/tijmenNL/348ee339daf7075b8e721f04a2c0f395

**Comment by lminiero** [comment]
Mh, that's weird, as the event handler support in the SIP plugin is indeed a bit of a hack, but to my knowledge only for outgoing messages. In fact, Sofia doesn't give us to the full outgoing messages: to overcome that, we have to basically override the logger (where full messages are printed) and extract them from there,

	if(notify_events && callback->events_is_enabled()) {
		JANUS_LOG(LOG_WARN, "sofia-sip logs are going to be redirected and they will not be shown in the process output\n");
		/* Enable the transport logging, as we want to have access to the SIP messages */
		setenv("TPORT_LOG", "1", 1);
		su_log_redirect(NULL, janus_sip_sofia_logger, NULL);
	}

The `janus_sip_sofia_logger` is where we do that, and if you had an issue there, I guess we could have blamed a broken parsing in our code. For incoming messages we should have access to the complete message, though, as we retrieve that any time the Sofia callback is invoked:

	/* Notify event handlers about the content of the whole incoming SIP message, if any */
	if(notify_events && gateway->events_is_enabled() && ssip) {
		/* Print the incoming message */
		size_t msg_size = 0;
		msg_t *msg = nua_current_request(nua);
		if(msg) {
			char *msg_str = msg_as_string(ssip->s_home, msg, NULL, 0, &msg_size);
			json_t *info = json_object();
			json_object_set_new(info, "event", json_string("sip-in"));
			json_object_set_new(info, "sip", json_string(msg_str));
			gateway->notify_event(&janus_sip_plugin, session->handle, info);
			su_free(ssip->s_home, msg_str);
		}
	}

No idea why it's broken for you, but it suggests an issue in either `nua_current_request` or `msg_as_string`, as we use Sofia APIs to build the message we need.

**Comment by lminiero** [comment]
Scratch that, I misunderstood the problem... I thought you were talking of event handlers, but this is an issue on an API event related to the content of an incoming SIP MESSAGE, so something completely different!

In that case, we're just accessing some Sofia struct fields to access the content:

```
	const char *content_type = sip->sip_content_type->c_type;
	char *payload = sip->sip_payload->pl_data;
```

If that's what you see in the message, it's likely that that's how Sofia parsed those messages. Do you have a Wireshark capture of those messages? We can try replicating the issue with sipp, to see if it happens here as well.

**Comment by tijmenNL** [comment]
I updated the gist with a ngrep and the corresponding janus events received over the websocket

**Comment by lminiero** [comment]
Tried replicating the issue with SIPp and I can't. Please try to prepare a SIPp script that consistently replicates the issue for you, and in case the same script causes the same issue to me I'll have a look.

**Comment by lminiero** [comment]
@tijmenNL any update on this?

**Comment by tijmenNL** [comment]
I tried replicating it but there was no definitive/reliable method to get it. I noticed that it happened after Janus was running for a while. 

Since then I moved the container where Janus was running to a faster filesystem. After that I did not see it (I think), but it is hard to say it will never happen, since I could not reliably reproduce it anyway.

So this can be closed. I will reopen the issue if it happens again and if I have more info.
